### ISSUE TITLE
`builds` trouble with `OmegaConf.to_object`

### ISSUE THREAD

**Issue by Jasha10** [mit-ll-responsible-ai/hydra-zen#242]
I'm having trouble when calling [`OmegaConf.to_object`](https://omegaconf.readthedocs.io/en/2.1_branch/usage.html#omegaconf-to-object) on DictConfig objects backed by builds-generated dataclasses. Here is an example showing the difference in behavior between a hand-coded dataclass and one generated using `builds`:
```python
# repro.py
from dataclasses import dataclass

from hydra_zen import builds
from omegaconf import OmegaConf


def foo(x: int) -> int:
    return x


@dataclass
class FooConfHandCoded:
    _target_: str = "__main__.foo"
    x: int = 10


FooConfBuilds = builds(foo, x=10)

foo_conf_hand_coded = OmegaConf.structured(FooConfHandCoded)
foo_conf_builds = OmegaConf.structured(FooConfBuilds)

print("\nUsing hand-coded dataclass:")
print(OmegaConf.to_object(foo_conf_hand_coded))

print("\nUsing builds:")
print(OmegaConf.to_object(foo_conf_builds))
```
```text
$ python repro.py

Using hand-coded dataclass:
FooConfHandCoded(_target_='__main__.foo', x=10)

Using builds:
Traceback (most recent call last):
  File "/home/jasha10/tmp/repro.py", line 27, in <module>
    print(OmegaConf.to_object(foo_conf_builds))
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/omegaconf/omegaconf.py", line 574, in to_object
    return OmegaConf.to_container(
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/omegaconf/omegaconf.py", line 553, in to_container
    return BaseContainer._to_content(
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/omegaconf/basecontainer.py", line 249, in _to_content
    return conf._to_object()
  File "/home/jasha10/miniconda3/envs/pysc/lib/python3.9/site-packages/omegaconf/dictconfig.py", line 752, in _to_object
    result = object_type(**field_items)
TypeError: __init__() got an unexpected keyword argument '_target_'
```

This fundamentally comes down to a difference in type signatures between `FooConfHandCoded.__init__` and `FooConfBuilds.__init__`:
```python
>>> import inspect
>>> inspect.signature(FooConfHandCoded)
<Signature (_target_: str = '__main__.foo', x: int = 10) -> None>
>>> inspect.signature(FooConfBuilds)
<Signature (x: int = 10) -> None>
```

**Comment by rsokl** [comment]
# TL;DR

- This is an issue with `OmegaConf.to_object`. The issue arises for a config backed by any structured config that possesses a field that is excluded from its signature, (for dataclasses, via `dataclasses.field(init=False)`).
- `builds` and `make_config` excludes all omegaconf/Hydra/zen-specific fields from the structured config's signature. I could add an option to toggle this behavior (e.g. `builds(..., zen_init=True)`). However, this particular issue is perhaps better addressed in omegaconf...
- There is a straightforward fix that will let `OmegaConf.to_object` work with structured configs that have non-init fields. Such structured configs already work with `instantiate` (i.e. non-init fields can already be overridden without issue). **I am happy to open a PR for this on omegaconf's end**.

# Actual Response

Thanks so much for looking into this @Jasha10 !

This is due to the fact that `builds` sets all hydra/omegaconf/zen-specific dataclass fields with `init=False` to exclude those options from the signature. 

`OmegaConf.to_object(...)` will throw an error on any conf backed by a dataclass that has fields which are excluded from its signature:

```python
# repro.py
from dataclasses import dataclass, field

from omegaconf import OmegaConf

@dataclass
class InitTrue:
    x: int = field(default=1, init=True)

@dataclass
class InitFalse:
    x: int = field(default=1, init=False)
    
Conf = OmegaConf.structured(InitTrue)

print("\nAll dataclass fields included in sig:")
print(OmegaConf.to_object(Conf))

Conf = OmegaConf.structured(InitFalse)

print("\nSome dataclass fields excluded from sig:")
print(OmegaConf.to_object(Conf))
```

```shell
$ python repro.py 

All dataclass fields included in sig:
InitTrue(x=1)

Some dataclass fields excluded from sig:
Traceback (most recent call last):
  File "repro.py", line 22, in <module>
    print(OmegaConf.to_object(Conf))
  File "C:\Users\Ryan Soklaski\Anaconda3\envs\hydra-zen\lib\site-packages\omegaconf\omegaconf.py", line 574, in to_object
    return OmegaConf.to_container(
  File "C:\Users\Ryan Soklaski\Anaconda3\envs\hydra-zen\lib\site-packages\omegaconf\omegaconf.py", line 553, in to_container
    return BaseContainer._to_content(
  File "C:\Users\Ryan Soklaski\Anaconda3\envs\hydra-zen\lib\site-packages\omegaconf\basecontainer.py", line 249, in _to_content
    return conf._to_object()
  File "C:\Users\Ryan Soklaski\Anaconda3\envs\hydra-zen\lib\site-packages\omegaconf\dictconfig.py", line 752, in _to_object
    result = object_type(**field_items)
TypeError: __init__() got an unexpected keyword argument 'x'
```

## Solutions

There are two solutions that I can think of here, which are not exclusive from one another.

### Changing behavior / exposing new behavior in `builds`:

`builds` could expose a new argument: `zen_init: bool = False`, which could be toggles to include *all* fields in the dataclass' signature:

```python
def f(x: int): ...

# Signature:
# InitFalse(x: int = 1)
InitFalse = builds(f, x = 1, zen_partial=True, zen_meta=dict(a=1), zen_init=False)  # default behavior

# Signature:
# InitTrue(
#     _target_: str = 'hydra_zen.funcs.zen_processing',
#     _zen_target: str = '__main__.f',
#     _zen_partial: bool = True,
#     _zen_exclude: Tuple[str, ...] = ('a',),
#     x: int = 1,
#     a: Any = 1,
# )
InitTrue = builds(f, x = 1, zen_partial=True, zen_meta=dict(a=1), zen_init=True)  
```

As you can see, it is generally *much* easier for a user to reason about the signature of the resulting config when it solely reflects elements of the target's signature. Otherwise, users need to know things about the implementation details of structured/targeted configs. This is why our current behavior is the way it is. 

There are other reasons why having `set(signature(builds(<target>))) <= set(signature(<target>))` is a very nice property to have, but this post is already too long as it stands.


I would be happy to support `zen_init=True`, if it is a generally useful feature. It would be a substantial amount of work, but it is doable. That being said, as far as this particular issue is concerned, I think a solution in omegaconf would be better...


### Accommodating `init=False`fields in omegaconf

I believe that `OmegaConf.to_object` can accommodate structured config fields for which `init=False` quite easily (for both dataclasses and `attr` classes). I would be happy to open a PR that adds support for this.

I should point out that non-init fields can still be overridden as you'd expect:

```python
@dataclass
class Conf:
    _target_: str = field(init=False, default="builtins.int")
```
```python
>>> instantiate(Conf)
0
>>> instantiate(Conf())
0

>>> instantiate(Conf, _target_="builtins.dict")
{}
>>> instantiate(Conf(), _target_="builtins.dict")
{}
```

and they seem to otherwise behave as-expected in omegaconf/Hydra. This is issue that you found is the first/only problem I have seen with non-init fields. Thus I think that supporting non-init fields in `OmegaConf.to_object` would be a nice enhancement!

#### What a solution would look like


In [`to_object`](https://github.com/omry/omegaconf/blob/5e73ee6a1c0f3a63fc420bf856750b80001a5644/omegaconf/dictconfig.py#L718), `field_items` should consist of fields that *only exist in the structured config's signature* -- `[field.name for field in dataclasses.fields(obj) if field.init]` for the case of dataclasses -- and all other fields should be gathered by `nonfield_items`. (perhaps they should actually be named `init_field_items` and `non_init_field_items`, respectively). 

For dataclasses, `field_items` is currently populated by [`omegaconf._utils.get_dataclass_field_names`](https://github.com/omry/omegaconf/blob/5e73ee6a1c0f3a63fc420bf856750b80001a5644/omegaconf/_utils.py#L324-L325), which accesses the names of *all* fields via: `[field.name for field in dataclasses.fields(obj)]`.

 Let me know what you think, and thanks again for taking the time to look into this and report this issue!

**Comment by rsokl** [comment]
Actually... I think `__post_init__` poses more insidious problems than I original though, regardless of `init=False` fields.

Consider the following behavior that currently occurs in OmegaConf:

```python
from dataclasses import dataclass, field

@dataclass
class A:
    a: float
    b: float
    def __post_init__(self) -> None:
        self.a = self.a ** 2
        self.b = self.b ** 2
```
```python
>>> cfg = OmegaConf.create(A(a=2, b=4))

>>> cfg
{'a': 4.0, 'b': 16.0}

>>> OmegaConf.to_object(cfg)
A(a=16.0, b=256.0)
```

The fact that this doesn't roundtrip makes me think think that, in all cases, `__post_init__`  should be circumvented by `to_object`. That is, in the case that `__post_init__` is present, the dataclass should be initialized and then *all* of its fields should be set via `setattr`!

**Comment by rsokl** [comment]
So, basically I think that `to_object` should do the following in the case where `__post_init__` is present:

```python
# to_object({"a": 2, "b": 4.0}) would do:
obj = A(a=2.0, b=4.0)
obj.a = 2.0
obj.b = 4.0
```

If a user wants to run `__post_init__`, they can just do it!

```python
>>> obj
A(a=2.0, b=4.0)

>>> obj.__post_init__()  # or perhaps this can be called via `to_object(cfg, post_init=True)` 
A(a=4.0, b=16.0)
```

**Comment by rsokl** [comment]
@Jasha10 if it sounds good to you, I'll close this and move the discussion to [omegaconf#789](https://github.com/omry/omegaconf/issues/789), where I'll summarize the relevant parts of this discussion (but *much* more concisely).

**Comment by Jasha10** [comment]
That would be great. Thanks!
